Port login page to Preact

mtomilov opened the pull request:
Refs https://github.com/hypothesis/h/issues/9588

Here we port login page to react in a hybrid way.
React app renders the form, but we actually submit it. And POST request is handled by the backend.
Login form controller is mostly based on group form controller.
For now it only affects regular login and open authorization popup login still uses Jinja2 for easier porting, which we'll address next.
Reusable components like text field will need to be lifted up from group form app.

Testing
===
- Go to `/login` page
- Try logging with valid and invalid credentials
- Go to `/login?for_oauth=true` page
- Try logging with valid and invalid credentials to make sure it continues to work as before

robertknight commented:
For the files which are re-used unchanged from the group forms app, I would suggest to simply re-import them from there for the moment. One of the frontend devs can then look at moving them to a shared directory and adjusting the tests etc.

mtomilov commented:
> For the files which are re-used unchanged from the group forms app, I would suggest to simply re-import them from there for the moment. One of the frontend devs can then look at moving them to a shared directory and adjusting the tests etc.

Thank you. I was going to lift them up `/scripts`, but sure not problem.
Duplicated them for draft because I wanted to make sure that I can reuse them as is.

mtomilov commented on h/static/scripts/group-forms/components/forms/TextField.tsx:
```diff
@@ -60,6 +74,9 @@ export type TextFieldProps = {
   classes?: string;
+
+  /** Optional error message for the field. */
+  fieldError?: string;
```
And this is field validation error we get from backend login schema.

mtomilov commented on h/views/accounts.py:
```diff
@@ -107,24 +109,63 @@ def __init__(self, request):
+        csrf_token = get_csrf_token(self.request)
+
+        return {
+            "styles": self.request.registry["assets_env"].urls("group_forms_css"),
```
Reusing group form styles because basically these are just tailwind styles, which is exactly the same we need for login form.

robertknight replied:
Something to know about how Tailwind works is that it scans a specified set of files looking for Tailwind classes that are used and only generates those styles in the CSS. The set of files that are scanned includes all of our TSX files under `h/static/scripts`, so re-using the CSS here will work. It will probably make sense to rename this bundle to something more general.

robertknight reviewed (commented):
This is looking good overall from an initial glance. I think we should remove the visible "required" markers from the fields in this form, though we will need to keep the `required` attribute on the form field. This will require a small change to the `TextField` API.

mtomilov commented on h/static/scripts/group-forms/components/forms/TextField.tsx:
```diff
@@ -99,17 +119,27 @@ export default function TextField({
         autocomplete="off"
         required={required}
+        {...(name ? { name: name } : {})}
+        {...(inputType === 'password' ? { type: 'password' } : {})}
```
Perfect, thank you!

mtomilov commented on h/static/scripts/login-forms/components/LoginForm.tsx:
```diff
@@ -0,0 +1,76 @@
+          <span>
+            {/* These are in a child span to avoid a gap between them. */}
+            <Star />
+            &nbsp;Required
```
Yep, should've made it the same as the original form rather than taking that from group form. Thanks.

mtomilov commented on h/static/styles/login-forms.css:
```diff
@@ -0,0 +1,6 @@
+/* CSS entry point for the login app. */
```
Yep, removed

mtomilov commented:
> This is looking good overall from an initial glance. I think we should remove the visible "required" markers from the fields in this form, though we will need to keep the `required` attribute on the form field. This will require a small change to the `TextField` API.

Added separate `showRequired` argument, defaulting to `required` which we set in `LoginForm`.
Probably we want opt-out behavior now for the indication here.

robertknight commented:
If you try to login with an un-activated account you see this error message which squashes the username label slightly:

<img width="562" alt="Account activation error" src="https://github.com/user-attachments/assets/fd943d8f-7d62-42ec-905e-23a49a8c274e" />

A simple solution will be to just move the error below the form field:

<img width="542" alt="Error below form field" src="https://github.com/user-attachments/assets/b9fd030c-adf4-4ff1-a8bf-b038b31fd552" />

```diff
diff --git a/h/static/scripts/group-forms/components/forms/TextField.tsx b/h/static/scripts/group-forms/components/forms/TextField.tsx
index 6d8c2f05b..a3ec6103d 100644
--- a/h/static/scripts/group-forms/components/forms/TextField.tsx
+++ b/h/static/scripts/group-forms/components/forms/TextField.tsx
@@ -114,12 +114,6 @@ export default function TextField({
     <div>
       <div className="flex items-center mb-1">
         <Label htmlFor={id} text={label} />
-        {fieldError && !hasCommitted && (
-          <>
-            <div className="grow" />
-            <ErrorNotice message={fieldError} />
-          </>
-        )}
       </div>
       <InputComponent
         id={id}
@@ -134,6 +128,11 @@ export default function TextField({
         name={name}
         type={inputType}
       />
+      {fieldError && !hasCommitted && (
+        <div className="my-2">
+          <ErrorNotice message={fieldError} />
+        </div>
+      )}
       {typeof maxLength === 'number' && (
         <CharacterCounter
           value={[...value].length}
```

robertknight commented on h/static/scripts/group-forms/components/forms/TextField.tsx:
```diff
@@ -99,7 +112,15 @@ export default function TextField({
-      <Label htmlFor={id} text={label} required={required} />
+      <div className="flex items-center mb-1">
+        <Label htmlFor={id} text={label} />
+        {fieldError && !hasCommitted && (
```
There is an accessibility issue here that the error notice is not connected to the form field. The `error` prop solves this by setting `aria-invalid` on the field and also a custom validity message. I agree with your comment about the current presentation of `error` not being suitable for these messages. I think it will be best to merge this as-is and then I can take a look at revising the `TextField` APIs to support these kinds of errors in a better way.

The group forms handle server errors by showing them at the bottom of the form. However this is becoming problematic as the form has grown longer, because it now shown a long way from the associated field.

<img width="654" alt="Group form validation error" src="https://github.com/user-attachments/assets/01bd50e1-aa48-4bc7-ab6a-8ca3cc57d482" />

robertknight commented on h/static/scripts/login-forms/components/LoginForm.tsx:
```diff
@@ -0,0 +1,70 @@
+          name="username"
+          value={username}
+          fieldError={config.formErrors?.username ?? ''}
+          onChangeValue={handleChangeUsername}
```
You can just pass `setUsername` here and `setPassword` for the field below and remove the `handle*` wrappers which don't do anything.

mtomilov replied:
Looks good, thank you!

robertknight commented on h/static/scripts/login-forms/components/LoginForm.tsx:
```diff
@@ -0,0 +1,70 @@
+        <div className="text-right">
+          <a
+            href={routes.forgotPassword}
+            className="text-grey-5 text-xs underline"
```
Can you change this to `text-sm` (14px). We're moving to standardizing on 14px as the standard base font size for readability.

mtomilov replied:
Sure, maybe it's worth changing the text size of the footer as well? Made it 14px inline for simplicity.

robertknight commented on h/static/scripts/login-forms/components/test/LoginForm-test.js:
```diff
@@ -0,0 +1,128 @@
+    } = elements;
+
+    assert.equal(form.prop('action'), routes.login);
+    assert.equal(form.prop('method'), 'POST');
```
We usually skip testing of static JSX (`method="POST"` here) and only test the dynamic parts.

robertknight commented on h/static/scripts/login-forms/routes.ts:
```diff
@@ -0,0 +1,11 @@
+  login: '/login',
+  forgotPassword: '/forgot-password',
+  signup: '/signup',
+};
```
We should combine this and `group-forms/routes.ts` in future since the routes have to be unique across all apps.

robertknight reviewed (commented):
Just a few small changes requested and then this is good to merge. I did note an accessibility issue with the validation error, but I think it would be best to avoid tackling that in this PR. I can file an issue and then look into it after this lands.

mtomilov commented:
> If you try to login with an un-activated account you see this error message which squashes the username label slightly:

Makes sense, moved it down but also moved character count up on label line because the count wouldn't fit if the message is present. Not strictly required because right now we use either character count or these external errors but I think it's worth having the ability to use both if needed.
It affects group forms, but hopefully it's not a big issue.

robertknight commented on h/templates/accounts/login.html.jinja2:
```diff
@@ -1,5 +1,40 @@
+    <h1 class="form-header">Log in</h1>
+    <div id="login-form"></div>
+    <footer class="form-footer">
+      <span style="font-size: 14px">
```
Can you set the style to `0.875rem` and add a comment that this aligns with Tailwind's `text-sm`. See https://tailwindcss.com/docs/font-size#basic-example.

robertknight reviewed (approved):
Can you put the character counter back below the form, and then show the error message below that. Otherwise this is good to merge.
